### ISSUE TITLE
Dataloader Hangs in multiprocessing of epoch ends early

### ISSUE THREAD

**Issue by gbarello-uipath** [huggingface/accelerate#2258]
### System Info

```Shell
- `Accelerate` version: 0.23.0
- Platform: Linux-6.2.0-1018-azure-x86_64-with-glibc2.35
- Python version: 3.11.6
- Numpy version: 1.25.2
- PyTorch version (GPU?): 2.1.0+cu121 (True)
- PyTorch XPU available: False
- PyTorch NPU available: False
- System RAM: 866.10 GB
- GPU type: NVIDIA A100 80GB PCIe
- `Accelerate` default config:
        - compute_environment: LOCAL_MACHINE
        - distributed_type: MULTI_GPU
        - mixed_precision: no
        - use_cpu: False
        - debug: False
        - num_processes: 2
        - machine_rank: 0
        - num_machines: 1
        - gpu_ids: 0,1
        - rdzv_backend: static
        - same_network: True
        - main_training_function: main
        - downcast_bf16: no
        - tpu_use_cluster: False
        - tpu_use_sudo: False
        - tpu_env: []
```

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [X] My own task or dataset (give details below)

### Reproduction

I am training a model in a scenario where sometimes the batches are "bad" and we skip them at runtime. The nature of our dataset and model are such that it is hard to only ever load in batches that are "good" so we are instead skipping them in the training loop. This works fine until one of the dataloaders finishes an "epoch" out of sync with the others (since by chance it skipped more batches). At that point, training hangs with one process waiting to retrieve a batch and other processes waiting to gather something (in the real training loop its trying to gather gradients, I presume).

A minimal script that reproduces this is below. I verified that it fails for 2 and 4 GPUs:

```
from accelerate import Accelerator

accelerator = Accelerator()

import torch
from torch.utils.data import Dataset, DataLoader

class test_dataset(Dataset):
    def __init__(self, length: int):
        super().__init__()
        self.length = length

    def __len__(self)->int:
        return self.length
    
    def __getitem__(self, idx):
        return torch.randn(10,10)
    
dataset = test_dataset(1000)
dataloader = DataLoader(dataset, batch_size = 2, shuffle = True)
dataloader = accelerator.prepare(dataloader)

iters = 0
while True:
    iters += 1
    print(f"starting iteration {iters} for {accelerator.process_index}")
    for i, inp in enumerate(dataloader):
        print(f"Batch {i} for {accelerator.process_index}")
        if accelerator.process_index == 0 and i%3 != 0:
            continue
        bigbatch = accelerator.gather(inp)
```

### Expected behavior

I was expecting that training would continue since at each point each process should (so I believe) be able to load another batch independent of the other processes status, and eventually sync up in the gather step.

**Comment by gbarello-uipath** [comment]
Ok I see that if I set shuffle=False it doesn't hang. I presume that the shuffling happens across process, which is good in general.
